[Draft] V-HACD: update to v4.1 release.

Anutrix opened the pull request:
Requesting review and advice on this:
- I am not sure about conversions between std::vector and Godot's vector as well as V-HACD's Vertex and Godot's Vector3.
- A set of parameters got removed, renamed and added to V-HACD. I commented out the removed and new params as they have defaults. I will remove before merge.
- Should I add new params to `convex_decomposition_function` or let them default?
- Need to see which patches are needed as everything has gone to a single file.

Tested on MRP from https://github.com/godotengine/godot/issues/78515 and it looked fine to me:
![image](https://github.com/godotengine/godot/assets/11459028/19d8fe14-4f1a-44d5-80cc-eae26de61644)
![image](https://github.com/godotengine/godot/assets/11459028/01f6c92f-d1c9-4e75-a833-149b4017c714)
![image](https://github.com/godotengine/godot/assets/11459028/ec8cfd90-1bfa-4c50-9342-ae4c1bb4945f)


I have very little experience with this part of Godot so further testing before merge is highly needed and appreciated.

Anutrix committed:
V-HACD: Update to v4.1 release.

Calinou commented:
I recommend testing the MRPs linked in https://github.com/godotengine/godot/issues/29971.

akien-mga commented on modules/vhacd/register_types.cpp:
```diff
@@ -32,24 +32,33 @@
 #include "scene/resources/mesh.h"
 
+#define ENABLE_VHACD_IMPLEMENTATION 1
 #include "thirdparty/vhacd/public/VHACD.h"
```
Since there's only one file, I would remove the `public/` folder and just have `thirdparty/vhacd/VHACD.h`.

Also, remember to update `thirdparty/README.md` with information on what commit is currently included, and what file(s) to copy on updates.

We also had a bunch of patch files which may no longer be relevant, if so they should be removed too.

Anutrix commented:
Updated `thirdparty\README.md` too.
Moved `thirdparty/vhacd/public/VHACD.h` to thirdparty/vhacd/VHACD.h.

Got rid of the patches with following reasoning:

0001-bullet-namespace.patch
 - This patch is too old(3.2; 2019) while V-HACD got major refactor.
 - Removed since it compiles fine on MinGW anyways. Do let me know if anyone needs me to additionally test specific configuration.

0002-fpermissive-fix.patch
 - Used Fedora 30(because mingw-gcc 8.4 is needed and Fedora 30 has gcc but not mingw64-gcc-c++ 8.4+) and target=editor.
 - `docker run -d -it fedora:31 /bin/bash`. `docker exec -it e26c2d11d869 /bin/bash`. Install git and other compiler things in 2 doc pages for cross-compiling.
 - `git clone -b vhacd-4.1-upgrade --single-branch https://github.com/Anutrix/godot.git`
 - `scons platform=windows target=editor use_static_cpp=yes -j`nproc` use_lto=no use_llvm=yes gdnative_wrapper=yes deprecated=no`
 - Removed since we confirmed that GH-27926 doesn't occur in the new version.

0003-fix-musl-build.patch
 - Removed since there's no `PTHREAD_MUTEX_RECURSIVE_NP`.

0004-fix-uwp-arm-build.patch
 - Removed since SSE intrinsics don't seem to be used in the new version. (Also, since isn't UWP dropped?).

0005-fix-scale-calculation.patch
 - Removed since the fix is already present upstream and in the new version too.

0006-fix-gcc13.patch
 - Removed since `stdint.h` is present in the new version too.
 - I'm little unsure of this since I am not sure why `cstdint` instead was used for V-HACD and `stdint.h` was used for openxr in same patch.

**TODO:**
 - Decide defaults/values for new parameters. How should we change the params? Won't there be breaking changes?
 - Testing for regressions(for example, see boss_robot_head_collision above).
 - Any help with these is appreciated.

akien-mga commented on thirdparty/README.md:
```diff
@@ -796,12 +796,12 @@ number and run the script.
+- From `include/`: `VHACD.h`
 - `LICENSE`
 
 Some downstream changes have been made and are identified by
```
Since all patches are removed, I think this is no longer relevant?

akien-mga commented:
There's a warning raised by Clang which we need to fix as we treat all warnings as errors:
```
/thirdparty/vhacd/VHACD.h:5698:9: error: private field 'count' is not used [-Werror,-Wunused-private-field]
    int count;
        ^
```

The line can just be removed. Would be good to also PR the fix upstream.

Anutrix commented:
> The line can just be removed. Would be good to also PR the fix upstream.

I had made a ticket for it upstream https://github.com/kmammou/v-hacd/issues/148. Maybe a PR would have been better. The response on that repo is slow.

Also, should I be remove and note it in a godot patch file? Or is this change too minor for it?

akien-mga commented:
> Also, should I be remove and note it in a godot patch file? Or is this change too minor for it?

I guess you can just comment it out without bothering with a patch file, and we'll notice it if we ever update and it hasn't been fixed upstream.

Anutrix commented:
@akien-mga I commented count but now it looks like it's used:
```
=====
In file included from modules\vhacd\register_types.cpp:36:
./thirdparty/vhacd/VHACD.h:5708:7: error: member initializer 'count' does not name a non-static data member or base class
 5708 |     , count(0)
      |       ^~~~~~~~
1 error generated.

=====
```
Removed it too.

Anutrix commented:
Small summary on removed old params based on V-HACD's Readme.md and comments header notice:
 - Removed m_alpha and m_beta since 'The tuning parameters 'm_alpha' and 'm_beta' are gone and no longer used'.
 - Removed m_oclAcceleration since 'All OpenCL related code removed'.
 - Removed m_planeDownsampling, m_convexhullDownsampling, m_pca, m_mode and m_convexhullApproximation since 'Old plane splitting code removed'. All these seem to be all directly or indirectly plane-splitting related.
 - Removed m_minVolumePerCH since it was used in SimplifyConvexHulls which is last step of ComputeACD, that was completely refactored to instead PerformConvexDecomposition recursively.

fire commented:
Is this being worked on currently? Wondering if this is being updated or is it salvagable.

Anutrix commented:
> Is this being worked on currently? Wondering if this is being updated or is it salvagable.

I got busy with some stuff. I still may not be able to work on this for a while.
The build from this branch was still running fine last I checked(simple rebase might be needed) so likely salvageable. Maybe someone with more experience can take over from here.

Things left to do:
- Decide defaults/values for new parameters. In turn, modification of convex_decomposition_function and respective changes.
- Testing for regressions(for example, see boss_robot_head_collision above).

fire commented:
I'm thinking of salvaging it, but I also don't have much bandwidth. How's your situation?
